accel/tcg/log_instr.c

```diff
@@ -1081,6 +1097,361 @@ void qemu_log_instr_extra(CPUArchState *env, const char *msg, ...)
+            } else if (is_long) {
+                g_string_append_printf(string, bounce_buf, (args++)->ulongv);
+            } else if (is_short) {
+                g_string_append_printf(string, bounce_buf, (args++)->ushortv);
+            } else {
+                g_string_append_printf(string, bounce_buf, (args++)->uintv);
+            }
+            format = false;
+            i = 0;
+            break;
+        case 'e':
+        case 'E':
+        case 'f':
+        case 'g':
+        case 'G':
+            if (is_long) {
+                g_string_append_printf(string, bounce_buf, (args++)->doublev);
+            } else {
+                g_string_append_printf(string, bounce_buf, (args++)->floatv);
+            }
+            format = false;
+            i = 0;
+            break;
+        case 's':
+        case 'p':
+            g_string_append_printf(string, bounce_buf, (args++)->ptrv);
+            format = false;
+            i = 0;
+            break;
+        case '%':
+            format = false;
+            break;
+        case 'h':
+            is_short = true;
+            break;
+        case 'l':
+            if (is_long) {
+                is_long_long = true;
+            }
+            is_long = true;
+            break;
+        default:
+            break;
+        }
+    }
+    if (i != 256) {
+        g_string_append_printf(string, bounce_buf);
+    }
+
+#pragma clang diagnostic pop
+}
+
+TCGv_i64 qemu_log_printf_valid_entries;
+
+#define QEMU_PRINTF_LOG_OFFSET                                                 \
+    ((offsetof(ArchCPU, parent_obj) - offsetof(ArchCPU, env)) +                \
+     offsetof(struct CPUState, log_state.qemu_log_printf_buf))
+
+void qemu_log_printf_create_globals(void)
+{
+    qemu_log_printf_valid_entries = tcg_global_mem_new_i64(
+        cpu_env,
+        QEMU_PRINTF_LOG_OFFSET + offsetof(qemu_log_printf_buf_t, valid_entries),
+        "log_valids");
+}
+
+void qemu_log_gen_printf(DisasContextBase *base, const char *qemu_format,
+                         const char *fmt, ...)
+{
+
+    if (!qemu_base_logging_enabled(base)) {
+        return;
+    }
+
+    va_list args;
+    va_start(args, fmt);
+
+    size_t ndx = base->printf_used_ptr++;
+    assert(
+        ndx < QEMU_LOG_PRINTF_BUF_DEPTH &&
+        "Increase QEMU_LOG_PRINTF_FLUSH_BARRIER or QEMU_LOG_PRINTF_BUF_DEPTH");
+
+    size_t offset = QEMU_PRINTF_LOG_OFFSET +
+                    (sizeof(qemu_log_arg_t) * (QEMU_LOG_PRINTF_ARG_MAX * ndx));
+
+    int nargs = 0;
+    char t;
+
+    TCGv_i64 temp64 = tcg_temp_new_i64();
+    TCGv_i32 temp32 = tcg_temp_new_i32();
+
+    /* Store the format string out. */
+    size_t fmt_offset = QEMU_PRINTF_LOG_OFFSET +
+                        offsetof(qemu_log_printf_buf_t, fmts) +
+                        (sizeof(const char *) * ndx);
+
+#if UINTPTR_MAX == UINT32_MAX
+    tcg_gen_movi_i32(temp32, (uintptr_t)fmt);
+    tcg_gen_st_i32(temp32, cpu_env, fmt_offset);
+#else
```

Comment:
Maybe we should add `#elif UINTPTR_MAX == UINT32_MAX #else #error "Are you really trying to build QEMU purecap???` here?

Reply:
tcg-target.h will already kill the build if that happens. Los of other UINTPTR_MAX == UINT32_MAX ... else ... throughout the codebase that would need fixing up for 128-bit pointers as well, so this can probably be like them and not handle a default case.

Reply:
Sounds good, no need to fix this.

Reply:
UINTPTR_MAX is PTRADDR_MAX for CHERI, it's not like SIZEOF :)